MH-12885 Capture Agent Access Management

krinnewitz opened the pull request:
This patch adds the possibility to configure which user (groups) have
access to which capture agents. Whenever a new capture agent is
registered, a new role is generated which a user needs to have when
changing scheduling information affecting the capture agent.

This includes:

- Editing schedule information of scheduled events
- Changing the workflow settings of scheduled events
- Scheduling new events
- Deleting scheduled events
- Deleting capture agents

The administrator as well as the organization administrator always has
access to all capture agents. For other users, access can be configured
by adding/removing the generated roles using the admin UI.

Here some user stories describing the reasoning behind this feature:

- As person responsible for scheduling, I do not want other users to mess up with capture agents I'm responsible for
- As a user, I want to be able to manage my recordings, including cutting of the video
- As a user, I want to see upcoming recordings ("scheduled events"), so that I know what will be recorded for me in case other persons are responsible for the actual recording
- As a institution, I might have several independent groups of people that are responsible for different sets of capture agents

Visually, this PR does not change much, but here are some impressions:

There are know three possibilities considering adding scheduled events:

1. The logged in user is system administrator (ROLE_ADMIN), tenant administrator (<Tenant Admin Role>) or has permissions to access all capture agents: Opencast will behave exactly the same as those roles have access to all capture agents

2. The logged in user is neither system administrator nor tenant administrator. He has only access rights to a subset of capture agents: The Add Event wizard looks still the same but the user can only select capture agents he has permissions to access

3. The logged in user has access to no capture agents at all: In that case, the Add Event Wizard has been adopted as this is the most common kind of users in Opencast setups that work with unprivileged users:

![screen shot 2018-06-11 at 12 23 39](https://user-images.githubusercontent.com/1590263/41226825-f3738a9e-6d72-11e8-99e5-1457abb365ab.png)

Users with no access to capture agents can see scheduled events based on the access policy, but they are not allowed to edit technical scheduling information:

![screen shot 2018-06-11 at 12 23 57](https://user-images.githubusercontent.com/1590263/41226987-904ce8e2-6d73-11e8-914c-2310ff3c015e.png)


This work is sponsored by SWITCH.

Commit:
MH-12885 Capture Agent Access Management

This patch adds the possibility to configure which user (groups) have
access to which capture agents. Whenever a new capture agent is
registered, a new role is generated which a user needs to have when
changing scheduling information affecting the capture agent.

This includes:

- Editing schedule information of scheduled events
- Changing the workflow settings of scheduled events
- Scheduling new events
- Deleting scheduled events
- Deleting capture agents

The administrator as well as the organization administrator always has
access to all capture agents. For other users, access can be configured
by adding/removing the generated roles using the admin UI.

Co-authored-by: Michael Stypa <michael@stypa.info>
Co-authored-by: Philipp Middendorf <middendorf@plapadoo.de>

krinnewitz commented:
I will rebase & extend this PR shortly since #231 is now merged

krinnewitz commented:
Done

staubesv commented on modules/admin-ui/src/main/resources/public/org/opencastproject/adminui/languages/lang-en_US.json:
```diff
@@ -130,6 +131,7 @@
      "USER_BLACKLIST_NOT_DELETED":  "The blacklist could not be deleted",
      "LOCATION_DELETED": "The location has been deleted",
      "LOCATION_NOT_DELETED": "The location could not be deleted",
+     "LOCATION_NOT_DELETED_NOT_AUTHORIZED": "The location could not be deleted, because you don't have the persmission to do so.",
```
Typo: per**s**mission

krinnewitz replied:
thanks, fixed

staubesv committed:
MH-12885 Added documentation for capture agent access control

staubesv commented:
Hi @gregorydlogan 

Thanks for taking the review! 

I've realised that this PR is currently missing the documentation and created a [PR](https://github.com/plapadoo/opencast/pull/18) for that.

@krinnewitz Can you please merge that PR?

staubesv committed:
MH-12885 Fixed typos in documentation

gregorydlogan commented:
I haven't had a chance to actually look at this code yet, but ideally the docs and code would be in the same PR.  Can we rebase the docs into this PR, or is that going to cause issues on your end?

Edit: Ah, that's not the opencast repo, so I can't merge it in locally :)  Nevermind!

staubesv commented:
@gregorydlogan The PR will be merged into this one. Currently waiting for @krinnewitz or @pmiddend to merge as I don't have write permissions for that repo, too.

staubesv commented:
Argh... damn!

@krinnewitz I've provided some fixes to the community which once again break one of your pull requests... sorry for that :-)

Can you please fix the merge conflicts?

krinnewitz commented:
ok, done @staubesv

Commit:
Fix edit scheduled events modal's "next" button

When the user doesn't have access to the capture agent of any of the
selected events, he shouldn't be able to continue.

staubesv committed:
Use correct JavaScript regex

This makes capture agent IDs containing non-alphanumerical characters work correctly.

gregorydlogan commented:
@krinnewitz, more conflicts.

gregorydlogan commented on docs/guides/user/docs/advanced/capture-agent-access.md:
```diff
@@ -0,0 +1,42 @@
+
+**Example**
+
+- Capture agent ID: `Opencast's Capture Agent`
```
IIRC neither spaces nor quotes are allowed, but I suppose I probably forgot to add enforcement to that.  Sigh...

krinnewitz replied:
I will just change the example to "OpencastsCaptureAgent"

gregorydlogan commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java:
```diff
@@ -337,8 +338,9 @@ public Response getEventResponse(@PathParam("eventId") String id) throws Excepti
                   @RestResponse(responseCode = HttpServletResponse.SC_NOT_FOUND, description = "The event could not be found."),
                   @RestResponse(responseCode = HttpServletResponse.SC_UNAUTHORIZED, description = "If the current user is not authorized to perform this action") })
-  public Response deleteEvent(@PathParam("eventId") String id) throws NotFoundException, UnauthorizedException {
+  public Response deleteEvent(@PathParam("eventId") String id) throws UnauthorizedException, SearchIndexException {
```
If we're getting rid of the NotFoundException we should also remove line 339.

krinnewitz replied:
Done

gregorydlogan commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java:
```diff
@@ -120,6 +120,7 @@
 import org.opencastproject.security.api.User;
 import org.opencastproject.security.urlsigning.exception.UrlSigningException;
 import org.opencastproject.security.urlsigning.service.UrlSigningService;
+import org.opencastproject.security.util.SecurityUtil;
```
I see SecurityUtil.checkAgentAccess, and checkAgentAccess in a bunch of places.  IMO, either always wrap, or never wrap the check.

krinnewitz replied:
Done

gregorydlogan commented on modules/admin-ui/src/main/webapp/scripts/modules/events/partials/eventActionsCell.html:
```diff
@@ -6,6 +6,7 @@
 </a>
 
 <!-- Remove button for published events. -->
+<!-- TODO: Falls event status SCHEDULED Berechtigungen prüfen. Ggf Button deaktivieren bzw nicht anzeigen. -->
```
Auf Englisch bitte :)

krinnewitz replied:
whoops, looks like I forgot to delete this comment :P

gregorydlogan commented on modules/index-service/src/main/java/org/opencastproject/index/service/api/IndexService.java:
```diff
@@ -138,13 +138,15 @@ Response updateGroup(String id, String name, String description, String roles, S
    *           Thrown if the provided request was inappropriate.
    */
-  String createEvent(HttpServletRequest request) throws IndexServiceException, IllegalArgumentException;
+  String createEvent(HttpServletRequest request, boolean checkAgentAccess) throws IndexServiceException, IllegalArgumentException;
```
Under what circumstances would we *not* want to check the access here?

krinnewitz replied:
Seems to not be used. Removed.

gregorydlogan reviewed (changes requested):
Looks good to me, will test tomorrow hopefully.

pmiddend commented:
I fixed the conflicts. Thanks @gregorydlogan for your review so far. plapadoo is on vacation this week, so there won't be any more fixes till Friday.

Commit:
Fix review findings

krinnewitz commented:
@gregorydlogan thanks for your review so far. I have resolved the issues you came up with. Feel free to continue the review :-)

gregorydlogan commented:
Looks great guys.  Sole complaint is that the CA is still visible there's no obvious reason why I couldn't schedule against it as a user who does not have permissions to use it.  Is this by design?  Either way, this seems like a reasonable PR to me!

krinnewitz commented:
Thanks @gregorydlogan. You mean the CA's you don't have write access to shouldn't show up in the "locations" area of the adminUI? I have no strong feelings about that. Maybe you still want to see that the CA exists and which status it has?

gregorydlogan commented:
@krinnewitz Yeah, fair enough.  This would be low impact even if the users aren't supposed to be able to see the agents, so let's see if we have complaints before we go to the work of fixing that :)
